### DISCUSSION TITLE
function definition in python

### DISCUSSION THREAD

**Discussion by michaelli66** [ryanorendorff/ryanorendorff.github.io#19] [General]
In your blog define a python function once and for all, you had a define_once function and you checked  if the function already exited by checking it's in the global dictionary like below:
```
def define_once(f):
    if f.__name__ in globals().keys():
          #


```
I am wondering:
-  if python functions with the same name but different argument  lists are considered the same function or not?
-  Is there a tool to check the names of a function's arg list?

**Comment by ryanorendorff** [comment]
Great questions!

> if python functions with the same name but different argument lists are considered the same function or not?

The AST includes the function arguments, so functions with the same name but different arguments are considered different. For example, the following function

```python
def f(a: int) -> int:
    return a + 5
```

generates the following AST string.

```
Module(
    body=[
        FunctionDef(
            name="f",
            args=arguments(
                posonlyargs=[],
                args=[arg(arg="a", annotation=Name(id="int", ctx=Load()))],
                kwonlyargs=[],
                kw_defaults=[],
                defaults=[],
            ),
            body=[
                Return(
                    value=BinOp(
                        left=Name(id="a", ctx=Load()), op=Add(), right=Constant(value=5)
                    )
                )
            ],
            decorator_list=[],
            returns=Name(id="int", ctx=Load()),
        )
    ],
    type_ignores=[],
)
```

So not only do the arguments matter, but all of the following matter: if the function arguments receive new annotations, have their order changed, made into positional only arguments, are given default values, etc.

> Is there a tool to check the names of a function's arg list?

The `inspect.signature` function will give you a function signature that you can parse through. You can see an example of parsing through function arguments [here](https://github.com/facebookresearch/labgraph/blob/f29eb520335e883bb5244a6e4400cf34a281ec8a/signal_processing/nodes/function_to_node.py#L351-L367).
